### ISSUE TITLE
Semantics of rebind_simd / resize_simd for cv/ref-qualified simd types?

### ISSUE THREAD

**Issue by rarutyun** [mattkretz/std-simd-feedback#37]
The issue that I have here is I don't understand where those support cv-qualifiers or references.

For example I would assume that if two expression `rebind_simd<simd<whatever, whatever>, to-x>` and `rebind_simd<const simd<whatever, whatever, to-x>` would give me the same result except `const` qualification (if it was for the original argument) for rebound `simd` type would be preserved.

Note: What I describe works for `std::allocator_traits<const A>::rebind_alloc<args>` today but drops `const` for resulted `A` and it doesn't work for the references (https://godbolt.org/z/Ms6MKjKq1). Probably we should do the same or define this semantics in other way. Currently it looks like the proposal doesn't support such a scenario.

**Comment by mattkretz** [comment]
Right now:

> The member `type` is present if and only if
• `V` is either `simd<U, Abi0>` or `simd_mask<U, Abi0>`, where `U` and `Abi0` are deduced from `V`, and
• `T` is a vectorizable type, and [...]

So cv- or ref-qualified `V` or `T` results in no `type` member.

Do you believe that's too restrictive?

**Comment by rarutyun** [comment]
> Right now:
> 
> > The member `type` is present if and only if
> > • `V` is either `simd<U, Abi0>` or `simd_mask<U, Abi0>`, where `U` and `Abi0` are deduced from `V`, and
> > • `T` is a vectorizable type, and [...]
> 
> So cv- or ref-qualified `V` or `T` results in no `type` member.

That's what I think off. My understanding was the same. Thanks for confirming.

> Do you believe that's too restrictive?

I do, yes. At least it should be the same as for std::allocator_traits, not sure if there are super important use-cases for ref-qualifiers.

**Comment by mattkretz** [comment]
What use case are you thinking of? `rebind_simd_t<int, decltype(v)>`? `v` could be `const` and/or `ref` qualified, no?

**Comment by rarutyun** [comment]
> What use case are you thinking of? `rebind_simd_t<int, decltype(v)>`? `v` could be `const` and/or `ref` qualified, no?

Yes, you are right. But I am wondering why `allocator_traits` don't support that.

As I said we can propose what we believe is right. The interesting questions there still would be what result type should rebind produce? I mean if original type was cv or ref qualified does this mean that result type should preserve the same qualifiers as the original one?

**Comment by mattkretz** [comment]
Again, my answer depends on use case. E.g.
```c++
void f(auto&& x) {
  auto y = static_cast<rebind_simd_t<int, decltype(x)>>(x);
```
Here you don't want to keep the ref-qualifiers. You might want to keep `const`, but why not be explicit about wanting `const` by writing `const auto y`. I mean this obviously creates a copy, why should it be `const`?

All that said, while I'm sympathetic to the use case, I see `rebind_simd` as a lower-level tool that should to the simple straightforward thing. I'd rather work on #34 instead.

**Comment by rarutyun** [comment]
> Again, my answer depends on use case. E.g.
> 
> ```c++
> void f(auto&& x) {
>   auto y = static_cast<rebind_simd_t<int, decltype(x)>>(x);
> ```
> 
> Here you don't want to keep the ref-qualifiers. You might want to keep `const`, but why not be explicit about wanting `const` by writing `const auto y`. I mean this obviously creates a copy, why should it be `const`?
> 
> All that said, while I'm sympathetic to the use case, I see `rebind_simd` as a lower-level tool that should to the simple straightforward thing. I'd rather work on #34 instead.

I think what you described is the most realistic use-case. I have thoughts about some pure template stuff when we operate on types but not on objects but those seem to be more advanced and less common.

As I said I don't have a strong preference where `const` should be preserved or not. With the use-case you provided it likely should not be. The thing I am mostly concerned is it should work.

If we look at the example you provided for the following call:

```cpp
const std::simd<float> s;
f(s);
```
`decltype(x)` is `const std::simd<float>&&`, which doesn't work today.

I am wondering why rebind for allocators doesn't work today with references but at least it works for *cv*-qualifiers. Probably we should go further with `std::simd`.

**Comment by mattkretz** [comment]
FWIW (https://godbolt.org/z/nj6n8fY5b):
```c++
void f0(std::make_unsigned_t<int>) {}
void f1(std::make_unsigned_t<const int>) {}
// ill-formed:
//void f2(std::make_unsigned_t<int&>) {}
//void f3(std::make_unsigned_t<const int&>) {}

void f(auto&& x) {
  auto y = static_cast<std::make_unsigned_t<decltype(x)>>(x);
}

// all ill-formed:
//template void f(int&);
//template void f(int&&);
//template void f(const int&);
//template void f(const int&&);
```
An argument can be made that `rebind_simd_t` should be consistent with `make_unsigned_t`.
